src/partialBind.ts

```diff
@@ -0,0 +1,76 @@
+/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/prefer-readonly-parameter-types, @typescript-eslint/no-unsafe-return --
+ * Function inference doesn't work when `unknown` is used as the parameters
+ * generic type, it **has** to be `any`.
+ */
+import type { IterableContainer, RemedaTypeError } from "./internal/types";
+
+type PartialBindError<Message extends string | number> = [
+  RemedaTypeError<"partialBind", Message>,
+];
+
+type RemovePrefix<
+  T extends IterableContainer,
+  Prefix extends IterableContainer,
+> = Prefix["length"] extends 0
+  ? T
+  : T["length"] extends 0
+    ? PartialBindError<"Too many args provided to function">
+    : T extends readonly [infer THead, ...infer TRest]
+      ? Prefix extends readonly [infer PrefixHead, ...infer PrefixRest]
+        ? // Both T and Prefix have a non-rest head.
+          PrefixHead extends THead
+          ? RemovePrefix<TRest, PrefixRest>
+          : PartialBindError<"Argument of the wrong type provided to function">
+        : Prefix extends ReadonlyArray<THead>
+          ? // T has a non-rest head, Prefix is an array.
+            // Prefix could possibly be empty, so this has to be THead?.
+            [THead?, ...RemovePrefix<TRest, Prefix>]
+          : // Prefix is e.g. [...Array<number>, string]. We can't do a
+            // type-level prefix removal, so we return an error.
+            PartialBindError<"Can't infer type of provided args">
```

Comment:
Should the error be that we can't *construct* the signature, not infer it, as it is being inferred properly, just that the types don't match?